go/vt/vttablet/tabletserver/gc/tablegc.go

```diff
@@ -210,101 +194,113 @@ func (collector *TableGC) Open() (err error) {
+
 	return nil
 }
 
 // Close frees resources
 func (collector *TableGC) Close() {
 	log.Infof("TableGC - started execution of Close. Acquiring initMutex lock")
-	collector.initMutex.Lock()
+	collector.stateMutex.Lock()
+	defer collector.stateMutex.Unlock()
 	log.Infof("TableGC - acquired lock")
-	defer collector.initMutex.Unlock()
-	if atomic.LoadInt64(&collector.isOpen) == 0 {
+	if collector.isOpen == 0 {
 		log.Infof("TableGC - no collector is open")
 		// not open
 		return
 	}
 
 	log.Info("TableGC: closing")
-	for _, t := range collector.tickers {
-		t.Suspend()
+	if collector.cancelOperation != nil {
+		collector.cancelOperation()
 	}
 	log.Infof("TableGC - closing pool")
 	collector.pool.Close()
 	atomic.StoreInt64(&collector.isOpen, 0)
 	log.Infof("TableGC - finished execution of Close")
 }
 
-// Operate is the main entry point for the table garbage collector operation and logic.
-func (collector *TableGC) Operate(ctx context.Context) {
+// operate is the main entry point for the table garbage collector operation and logic.
+func (collector *TableGC) operate(ctx context.Context) {
 
-	addTicker := func(d time.Duration) *timer.SuspendableTicker {
-		collector.initMutex.Lock()
-		defer collector.initMutex.Unlock()
+	dropTablesChan := make(chan string)
+	purgeRequestsChan := make(chan bool)
+	transitionRequestsChan := make(chan *transitionRequest)
 
-		t := timer.NewSuspendableTicker(d, true)
-		collector.tickers = append(collector.tickers, t)
+	tickers := [](*timer.SuspendableTicker){}
+	addTicker := func(d time.Duration) *timer.SuspendableTicker {
+		t := timer.NewSuspendableTicker(d, false)
+		tickers = append(tickers, t)
 		return t
 	}
 	tableCheckTicker := addTicker(checkInterval)
-	leaderCheckTicker := addTicker(leaderCheckInterval)
 	purgeReentranceTicker := addTicker(purgeReentranceInterval)
 
+	for _, t := range tickers {
+		defer t.Stop()
+		// since we just started the tickers now, speed up the ticks by forcing an immediate tick
+		go t.TickNow()
+	}
+
 	log.Info("TableGC: operating")
 	for {
 		select {
-		case <-leaderCheckTicker.C:
-			{
-				func() {
-					collector.initMutex.Lock()
-					defer collector.initMutex.Unlock()
-					// sparse
-					shouldBePrimary := false
-					if atomic.LoadInt64(&collector.isOpen) > 0 {
-						if collector.tabletTypeFunc() == topodatapb.TabletType_PRIMARY {
-							shouldBePrimary = true
-						}
-					}
-
-					if collector.isPrimary == 0 && shouldBePrimary {
-						log.Infof("TableGC: transition into primary")
-					}
-					if collector.isPrimary > 0 && !shouldBePrimary {
-						log.Infof("TableGC: transition out of primary")
-					}
-
-					if shouldBePrimary {
-						atomic.StoreInt64(&collector.isPrimary, 1)
-					} else {
-						atomic.StoreInt64(&collector.isPrimary, 0)
-					}
-				}()
-			}
+		case <-ctx.Done():
+			log.Info("TableGC: done operating")
+			return
 		case <-tableCheckTicker.C:
 			{
-				_ = collector.checkTables(ctx)
+				log.Info("TableGC: tableCheckTicker")
```

Comment:
Won't this log every 5 seconds on a primary tablet?

Reply:
Same comments for the other logs on tickers

Reply:
> Won't this log every 5 seconds on a primary tablet?

The default interval is `1h`, and I see no reason why in production it would be any less then several minutes. In testing, where we set it to `5s` - yes. But then, do we care?

Reply:
I removed the log entry for `log.Info("TableGC: purgeReentranceTicker")`, which runs once per `1m`. Still very spacious IMO and shouldn't be a problem logging once per minute, but now removed. The rest of logs are either once per hour, or upon something actually being done.